### PR TITLE
Add DI for Main controller, providers and processors

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

On browsing resources
Capture d’écran de 2023-12-15 16-02-41

On creating a new resource
Capture d’écran de 2023-12-15 16-03-18

On invalid creation
Capture d’écran de 2023-12-15 16-04-16

On deleting a single resource
Capture d’écran de 2023-12-15 16-05-01

On delete multiple resources
Capture d’écran de 2023-12-15 16-00-59

### CODE REVIEW

File: src/Component/src/State/Processor/RespondProcessor.php [hunk 1]
```diff
@@ -23,19 +23,17 @@
...
 
     public function process(mixed $data, Operation $operation, Context $context): mixed
     {
-        $data = $this->decorated->process($data, $operation, $context);
-
         if ($data instanceof Response) {
             return $data;
         }
 
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/RespondProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/WriteProcessor.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        if (
+            $data instanceof Response ||
+            !($operation->canWrite() ?? true) ||
+            !$operation->getProcessor()
+        ) {
+            return $this->processor->process($data, $operation, $context);
+        }
+
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/WriteProcessor.php]:
```suggestion
```

File: src/Component/src/State/Provider/EventDispatcherProvider.php [hunk 3]
```diff
@@ -33,13 +33,16 @@ public function __construct(
 
     public function provide(Operation $operation, Context $context): object|array|null
     {
+        $data = $this->decorated->provide($operation, $context);
+
         if (
             $operation instanceof CollectionOperationInterface ||
             $operation instanceof ShowOperationInterface
         ) {
+            dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Provider/EventDispatcherProvider.php]:
```suggestion
```

File: src/Component/src/State/Provider/ReadProvider.php [hunk 4]
```diff
@@ -38,6 +38,7 @@ public function provide(Operation $operation, Context $context): object|array|nu
             return null;
         }
 
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Provider/ReadProvider.php]:
```suggestion
```

File: src/Component/src/Symfony/Request/State/TwigResponder.php [hunk 5]
```diff
@@ -47,6 +47,8 @@ public function respond(mixed $data, Operation $operation, Context $context): ?R
             return null;
         }
 
+        dump('Data');
+
```
Comment (Prometee) [inline src/Component/src/Symfony/Request/State/TwigResponder.php]:
```suggestion
```

File: src/Component/src/Doctrine/Common/State/RemoveProcessor.php [hunk 6]
```diff
@@ -34,6 +34,7 @@ public function process(mixed $data, Operation $operation, Context $context): mi
             return null;
         }
 
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/Doctrine/Common/State/RemoveProcessor.php]:
```suggestion
```

File: src/Component/src/Doctrine/Common/State/PersistProcessor.php [hunk 7]
```diff
@@ -40,6 +40,7 @@ public function process(mixed $data, Operation $operation, Context $context): mi
             $manager->persist($data);
         }
 
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/Doctrine/Common/State/PersistProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/EventDispatcherProcessor.php [hunk 8]
```diff
@@ -38,6 +38,7 @@ public function __construct(
      */
     public function process(mixed $data, Operation $operation, Context $context): mixed
     {
+        dump('PRE ' . __METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/EventDispatcherProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/EventDispatcherProcessor.php [hunk 9]
```diff
@@ -52,6 +53,7 @@ public function process(mixed $data, Operation $operation, Context $context): mi
 
         $result = $this->decorated->process($data, $operation, $context);
 
+        dump('POST ' . __METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/EventDispatcherProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/FlashProcessor.php [hunk 10]
```diff
@@ -23,30 +23,33 @@
...
             $request->isMethodSafe() ||
-            !$request->attributes->getBoolean('is_valid', true)
+            $format !== 'html' ||
+            !($operation->canWrite() ?? true)
         ) {
-            return $data;
+            return $this->processor->process($data, $operation, $context);
         }
 
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/FlashProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/PostEventDispatcherProcessor.php [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(mixed $data, Operation $operation, Context $context): mixed
+    {
+        $data = $this->processor->process($data, $operation, $context);
+
+        dump('POST ' . __METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/PostEventDispatcherProcessor.php]:
```suggestion
```

File: src/Component/src/State/Processor/PreEventDispatcherProcessor.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(mixed $data, Operation $operation, Context $context): mixed
+    {
+        $operationEvent = $this->operationEventDispatcher->dispatchPreEvent($data, $operation, $context);
+
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/State/Processor/PreEventDispatcherProcessor.php]:
```suggestion
```

File: src/Component/src/Symfony/Serializer/State/SerializeProcessor.php [hunk 13]
```diff
@@ -47,13 +45,17 @@ public function process(mixed $data, Operation $operation, Context $context): mi
...
         }
 
         if (null === $this->serializer) {
             throw new \LogicException(sprintf('You can not use the "%s" format if the Serializer is not available. Try running "composer require symfony/serializer".', $format));
         }
 
-        return $this->serializer->serialize($data, $format, $operation->getNormalizationContext() ?? []);
+        $serialized = $this->serializer->serialize($data, $format, $operation->getNormalizationContext() ?? []);
+
+        dump(__METHOD__);
```
Comment (Prometee) [inline src/Component/src/Symfony/Serializer/State/SerializeProcessor.php]:
```suggestion
```

**Comment by loic425** [comment]
Thx @Prometee That's a draft and I need that to debug ;) Of course I will remove these dump methods when it will be ready to review, but I think I will split that into few smaller PRs.

File: src/Bundle/Resources/config/services/controller.xml [hunk 14]
```diff
@@ -15,6 +15,14 @@
     <services>
         <defaults public="true" />
 
+        <service id="sylius.controller.main" class="Sylius\Resource\Symfony\Controller\MainController" public="false">
```
Comment (diimpp) [inline src/Bundle/Resources/config/services/controller.xml]: Naming scheme conflicts with possible sylius.resource `main`, like

> sylius.model.main.class
> sylius.controller.main
> sylius.repository.main

File: src/Component/src/Symfony/Controller/MainController.php [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        private HttpOperationInitiatorInterface $operationInitiator,
+        private RequestContextInitiatorInterface $requestContextInitiator,
+        private ProviderInterface $provider,
+        private ProcessorInterface $processor,
+    ) {
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        $operation = $this->operationInitiator->initializeOperation($request);
```
Comment (diimpp) [inline src/Component/src/Symfony/Controller/MainController.php]: On side note,

image

```patch
---         /** @var string|null $operationName */
---        $operationName = $request->attributes->get('_route');
+++        $operationName = $request->attributes->getString('_route', null);
```

File: src/Component/src/Symfony/Controller/MainController.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+        if (null === $operation) {
+            throw new \RuntimeException('Operation should not be null.');
```
Comment (diimpp) [inline src/Component/src/Symfony/Controller/MainController.php]: Feels like it should be a resource specific exception here, like
> InvalidOperationException extends \RuntimeException implements \Sylius\Resource\Exception\ExceptionInterface extends \Throwable

File: src/Bundle/Resources/config/services/controller.xml [hunk 17]
```diff
@@ -15,6 +15,14 @@
     <services>
         <defaults public="true" />
 
+        <service id="sylius.symfony.main_controller" class="Sylius\Resource\Symfony\Controller\MainController" public="false">
```
Comment (lchrusciel) [inline src/Bundle/Resources/config/services/controller.xml]: I would go with:
```suggestion
        <service id="sylius.main_controller" class="Sylius\Resource\Symfony\Controller\MainController" public="false">
```
or
```suggestion
        <service id="sylius.controller" class="Sylius\Resource\Symfony\Controller\MainController" public="false">
```

that should be enough

Comment (diimpp) [reply src/Bundle/Resources/config/services/controller.xml]: :+1:  for `sylius.main_controller`. Service definition already in SF context, so no need to specify it explicitly.

File: tests/Application/templates/layout.html.twig [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>{% block title %}Sylius{% endblock %}</title>
+    <link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 128 128%22><text y=%221.2em%22 font-size=%2296%22>⚫️</text></svg>">
```
Comment (diimpp) [inline tests/Application/templates/layout.html.twig:6]: Just out of curiosity, Is `data:image` just a black dot?

File: src/Bundle/Resources/config/services/state.xml [hunk 19]
```diff
@@ -12,65 +12,29 @@
...
+    </imports>
+
     <services>
         <service id="sylius.resource_factory.expression_language" class="Symfony\Component\ExpressionLanguage\ExpressionLanguage" />
 
-        <service id="sylius.state.provider" class="Sylius\Resource\State\Provider">
+        <service id="sylius.state_provider.locator" class="Sylius\Resource\State\Provider">
             <argument type="tagged_locator" tag="sylius.state_provider" />
         </service>
         <service id="Sylius\Resource\State\ProviderInterface" alias="sylius.state.provider" />
```
Comment (diimpp) [inline src/Bundle/Resources/config/services/state.xml]: Should this alias be updated to `sylius.state_provider` scheme?
```suggestion
        <service id="Sylius\Resource\State\ProviderInterface" alias="sylius.state_provider" />
```

File: src/Bundle/Resources/config/services/state.xml [hunk 20]
```diff
@@ -12,65 +12,29 @@
...
-        <service id="sylius.state.provider.dispatch_post_read_event"
-                 class="Sylius\Resource\Symfony\EventDispatcher\State\DispatchPostReadEventProvider"
-                 decorates="sylius.state.provider"
-                 decoration-priority="-32"
-        >
-            <argument type="service" id=".inner" />
-            <argument type="service" id="sylius.dispatcher.operation" />
-        </service>
-
         <service id="sylius.state.factory" class="Sylius\Resource\State\Factory">
```
Comment (diimpp) [inline src/Bundle/Resources/config/services/state.xml]: Should new service naming scheme be applied as well?
```suggestion
        <service id="sylius.state_factory" class="Sylius\Resource\State\Factory">
```

File: tests/Application/templates/layout.html.twig (hunk 18 above)
Comment (loic425) [reply tests/Application/templates/layout.html.twig:6]: To be honest, I don't remember where this layout come from. But yes, this seems to be a black dot.
Capture d’écran de 2024-01-16 08-45-04

File: src/Bundle/Resources/config/services/state/processor/write.xml [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+        </service>
+
+        <!-- Post-write-->
+        <service id="sylius.state_processor.dispatch_post_write_event" class="Sylius\Resource\Symfony\EventDispatcher\State\DispatchPostWriteEventProcessor" decorates="sylius.state_processor.locator" decoration-priority="200">
+            <argument type="service" id=".inner" />
+            <argument type="service" id="sylius.dispatcher.operation" />
+            <argument type="service" id="sylius.event_handler.operation" />
+        </service>
+
+        <service id="sylius.state.processor.bulk_aware" class="Sylius\Resource\State\Processor\BulkAwareProcessor" decorates="sylius.state_processor.locator" decoration-priority="100">
```
Comment (diimpp) [inline src/Bundle/Resources/config/services/state/processor/write.xml]:
```suggestion
        <service id="sylius.state_processor.bulk_aware" class="Sylius\Resource\State\Processor\BulkAwareProcessor" decorates="sylius.state_processor.locator" decoration-priority="100">
```

File: src/Bundle/Resources/config/services/state.xml [hunk 22]
```diff
@@ -12,65 +12,29 @@
...
-            <argument type="service" id="sylius.dispatcher.operation" />
-        </service>
+        <service id="Sylius\Resource\State\ProviderInterface" alias="sylius.state_provider" />
 
-        <service id="sylius.state.factory" class="Sylius\Resource\State\Factory">
+        <service id="sylius.state_factory" class="Sylius\Resource\State\Factory">
             <argument type="tagged_locator" tag="sylius.resource_factory" />
             <argument type="service" id="sylius.expression_language.argument_parser.factory" />
         </service>
         <service id="Sylius\Resource\State\FactoryInterface" alias="sylius.state.factory" />
```
Comment (diimpp) [inline src/Bundle/Resources/config/services/state.xml]:
```suggestion
        <service id="Sylius\Resource\State\FactoryInterface" alias="sylius.state_factory" />
```

**Review by diimpp — APPROVED** [review]

File: src/Bundle/Resources/config/services/state.xml [hunk 23]
```diff
@@ -12,65 +12,29 @@
 -->
 
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <imports>
+        <import resource="state/processor.xml" />
+        <import resource="state/provider.xml" />
+    </imports>
```
Comment (GSadee) [inline src/Bundle/Resources/config/services/state.xml]: You can put the generic import in the `services.xml` file
```
<imports>
    <import resource="services/**/*.xml" />
</imports>
```

File: src/Bundle/Resources/config/services/state.xml [hunk 24]
```diff
@@ -12,65 +12,29 @@
...
+
     <services>
         <service id="sylius.resource_factory.expression_language" class="Symfony\Component\ExpressionLanguage\ExpressionLanguage" />
 
-        <service id="sylius.state.provider" class="Sylius\Resource\State\Provider">
+        <service id="sylius.state_provider.locator" class="Sylius\Resource\State\Provider">
```
Comment (GSadee) [inline src/Bundle/Resources/config/services/state.xml:24]: Is this `.locator` suffix needed?
```suggestion
        <service id="sylius.state_provider" class="Sylius\Resource\State\Provider">
```

**Review by GSadee — APPROVED** [review]

File: src/Bundle/Resources/config/services/state.xml (hunk 24 above)
Comment (loic425) [reply src/Bundle/Resources/config/services/state.xml:24]: this is not really the generic one. the main one is the sylius.state_provider.read which is aliased as sylius.state_provider.main

File: src/Bundle/Resources/config/services/state.xml (hunk 23 above)
Comment (loic425) [reply src/Bundle/Resources/config/services/state.xml]: you're right, done (with a few change on your proposal).

File: src/Bundle/Resources/config/services/state.xml [hunk 25]
```diff
@@ -12,65 +12,30 @@
 -->
 
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <imports>
+        <import resource="state/processor.xml" />
+        <import resource="state/provider.xml" />
+        <import resource="state/**/*.xml" />
+    </imports>
```
Comment (GSadee) [inline src/Bundle/Resources/config/services/state.xml:19]: This is a small thing, but I suggested removing these imports and add a generic one to `services.xml`

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Loïc! :tada:

**Pull request merged**